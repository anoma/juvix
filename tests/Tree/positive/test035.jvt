type list {
  nil : list;
  cons : (*, list) â†’ list;
}

function mklst(integer) : list;
function mklst2(integer) : list;
function append(list, list) : list;
function flatten(list) : list;
function main() : *;

function mklst(integer) : list {
  br(eq(arg[0], 0)) {
    true: alloc[nil]()
    false: alloc[cons](arg[0], call[mklst](sub(arg[0], 1)))
  }
}

function mklst2(integer) : list {
  br(eq(arg[0], 0)) {
    true: alloc[nil]()
    false: alloc[cons](call[mklst](arg[0]), call[mklst2](sub(arg[0], 1)))
  }
}

function append(list, list) : list {
  case[list](arg[0]) {
    nil: arg[1]
    cons: alloc[cons](arg[0].cons[0], call[append](arg[0].cons[1], arg[1]))
  }
}

function flatten(list) : list {
  case[list](arg[0]) {
    nil: save {
      tmp[0]
    }
    cons: call[append](arg[0].cons[0], call[flatten](arg[0].cons[1]))
  }
}

function main() : * {
  seq(save(call[flatten](save(call[mklst2](4)) {
    seq(trace(tmp[0]), tmp[0])
  })) {
    seq(trace(tmp[0]), tmp[0])
  }, void)
}
