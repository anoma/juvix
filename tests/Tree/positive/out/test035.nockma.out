[[4 3 2 1 nil] [3 2 1 nil] [2 1 nil] [1 nil] nil]
[4 3 2 1 3 2 1 2 1 1 nil]
